Holodeck Fixes

SuperSayu opened the pull request:
Adds NODECONSTRUCT flag to structures and machines.  This is specifically used by the holodeck for things that should be immune to disassembly.  Fixes #13183.
Fixes #13277 laser beams spawning when the holdout program ends

Fixes invisible windoors in the holodeck.
Standardizes the fire axe's interaction with windows.
Fixes a runtime when holodeck tiles get broken or burned.

SuperSayu committed:
Sets up the the ABSTRACT flag to be used by machines and structures that should not be disassembled or break into parts.

This is specifically used by the holodeck for things that should be immune to disassembly.

Fixes invisible windoors in the holodeck.
Standardizes the fire axe's interaction with windows.
Fixes a runtime with holodeck tiles getting broken or burned.

tkdrg commented:
:+1: it will be good to have a systems in place for this kind of thing

Anonus commented on code/game/objects/items/weapons/twohanded.dm:
```diff
@@ -177,17 +177,12 @@
-		qdel(A)
+			W.spawnfragments() // this will qdel and spawn shards
+		else if(istype(A,/obj/structure/grille))
+			qdel(A)
```
hit the grille for like 90 or something, so it spawns rods

Aranclanos commented:
so... all of these checks just for the holodeck?

SuperSayu commented:
They can also be used on away missions or on space derelict stuff that
should not be moved to the station.

The holodeck has been part of tgcode for a long time.  With the upgrades I
added it is more feasible to add holodeck programs with real stuff in
them.  If you feel this code is going to waste, add more interesting
holodeck stuff, or extra holodecks that do different things.  I could give
you examples off the top of my head--I did in my holodeck PR already.

Also if you can think of a more reasonable way I am happy to oblige.  These
checks only happen in specific attackbys, they won't slow anything down.
On Nov 27, 2015 4:44 AM, "Aranclanos" <notifications@github.com> wrote:

> so... all of these checks just for the holodeck?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/13253#issuecomment-160097434>
> .
>

SuperSayu committed:
Better fireaxe-grill interactions

SuperSayu committed:
Fixes #13277 Lasers beams spawning after the holdout sim ends

Aranclanos commented:
I would probably just straight make new types for the holodeck instead of this.
like some esword rack that only accepts eswords to be stored, instead of a table that needs continuous fixing over the years

SuperSayu commented:
I disagree largely because that puts the onus on mappers to code variants
of anything they want to use.  Figuring out what needs to change when
something is holographic or map-special means diving pretty hard into the
code for each object mapped.  This actively discourages mappers from making
anything with machines in it, and that's not a good thing.

The machinery deconstruct abstract checks cut out a lot of that but not
everything. It does provide a framework of how those changes could our
should be done when they are added to something new.  And using the same
flag across structures and machines is good, although I half-want to go
back and make it is own flag name that just overlaps with ABSTRACT
numerically so the flag if clearer.  If rather not though.
On Nov 28, 2015 9:53 PM, "Aranclanos" <notifications@github.com> wrote:

> I would probably just straight make new types for the holodeck instead of
> this.
> like some esword rack that only accepts eswords to be stored, instead of a
> table that needs continuous fixing over the years
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/13253#issuecomment-160362347>
> .
>

SuperSayu committed:
Fixing conflicts:
	code/game/machinery/computer/computer.dm
	code/game/machinery/doors/windowdoor.dm
	code/game/objects/structures/window.dm

Cheridan commented:
aran that's a really horrible way to go about it (and the way it used to be done)

basically if you wanted to have a table in the holodeck, you had to make a snowflake holotable without deconstruction code; repeat for, like, any structure you wanted to use ever.

that said i don't know if i like the ABSTRACT flag for this, since its original purpose was to simulate abstract types from other languages, items that weren't ever meant to be spawned in-game.

SuperSayu commented:
Alright well since I have to fix conflicts again I will change the flag name and probably do minimal Z2 mapfixen

Razzem frazzem slow merge

SuperSayu committed:
Fix Conflicts:
	code/game/machinery/doors/windowdoor.dm

SuperSayu committed:
Changes flag name from ABSTRACT to NODECONSTRUCT

Removes holo-tables and holo-windows since those typepaths are no longer needed.
Updates Z2 to reflect this.  Removes spurious tag tags in Z2.
